### PR TITLE
Add repr(transparent) attribute to JavaVM struct

### PR DESCRIPTION
**Pull request by aleksuss**
## Overview

Adding of the attribute makes rust 1.46 compiler happy about issue: `extern` fn uses type `jni::JavaVM`, which is not FFI-safe`
 
```
warning: `extern` fn uses type `jni::JavaVM`, which is not FFI-safe
  --> src/utils/jni_cache.rs:67:39
   |
67 | pub extern "system" fn JNI_OnLoad(vm: JavaVM, _: *mut c_void) -> jint {
   |                                       ^^^^^^ not FFI-safe
   |
   = note: `#[warn(improper_ctypes_definitions)]` on by default
   = help: consider adding a `#[repr(C)]` or `#[repr(transparent)]` attribute to this struct
   = note: this struct has unspecified layout
``` 

### Definition of Done

- [x] There are no TODOs left in the code
- [ ] Change is covered by automated [tests](https://github.com/jni-rs/jni-rs/blob/master/CONTRIBUTING.md#tests)
- [ ] The [coding guidelines](https://github.com/jni-rs/jni-rs/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [ ] Public API has documentation
- [x] This change is not breaking **or** mentioned in the Changelog
- [ ] The [continuous integration build](https://www.travis-ci.org/jni-rs/jni-rs) passes

### CODE REVIEW

File: tests/threads_nested_attach_daemon.rs [hunk 1]
```diff
@@ -45,7 +45,5 @@ pub fn nested_attaches_should_not_detach_daemon_thread() {
         assert_eq!(jvm().threads_attached(), 1);
     }
     assert_eq!(jvm().threads_attached(), 1);
-
-    drop(env);
```
Comment (maurolacy) [inline tests/threads_nested_attach_daemon.rs:49]: Why are you removing this?

File: tests/threads_nested_attach_permanently.rs [hunk 2]
```diff
@@ -47,7 +47,6 @@ pub fn nested_attaches_should_not_detach_permanent_thread() {
     assert_eq!(jvm().threads_attached(), 1);
 
     // Dropping JNIEnv obtained from a permanent attach has no effect on the thread status.
-    drop(env);
```
Comment (maurolacy) [inline tests/threads_nested_attach_permanently.rs:50]: Same here.

File: src/wrapper/java_vm/vm.rs [hunk 3]
```diff
@@ -130,6 +130,7 @@ use crate::InitArgs;
 /// [actp]: struct.JavaVM.html#method.attach_current_thread_permanently
 /// [actd]: struct.JavaVM.html#method.attach_current_thread_as_daemon
 /// [spec-references]: https://docs.oracle.com/en/java/javase/12/docs/specs/jni/design.html#referencing-java-objects
+#[repr(transparent)]
```
Comment (maurolacy) [inline src/wrapper/java_vm/vm.rs:133]: Makes sense. Doc for reviewers: https://doc.rust-lang.org/1.26.2/unstable-book/language-features/repr-transparent.html

File: tests/threads_nested_attach_daemon.rs (hunk 1 above)
Comment (aleksuss) [reply tests/threads_nested_attach_daemon.rs:49]: Because of https://travis-ci.org/github/jni-rs/jni-rs/jobs/729379085#L258

Comment (aleksuss) [reply tests/threads_nested_attach_daemon.rs:49]: By the way. What's a point to call drop here ? As I can see `JNIEnv` doesn't implement `Drop` trait.

Comment (maurolacy) [reply tests/threads_nested_attach_daemon.rs:49]: I see. Thanks for the clarification.

I'm thinking this is probably related to a recent change (#255), where `JNIEnv` now implements `Copy`.

File: tests/threads_nested_attach_permanently.rs (hunk 2 above)
Comment (stanislav-tkach) [reply tests/threads_nested_attach_permanently.rs:50]: It looks like a part of the test, but `drop` for `JNIEnv` doesn't do anything. We have several similar tests, but only one of them (`nested_attaches_should_not_detach_guarded_thread`) uses `AttachGuard`.

**Review by stanislav-tkach — APPROVED** [review]

File: tests/threads_nested_attach_permanently.rs (hunk 2 above)
Comment (dmitry-timofeev) [reply tests/threads_nested_attach_permanently.rs:50]: Doesn't it make sense to leave it? It seems to verify that whatever `Drop` implementation JNIEnv has, it does not detach a permanently attached thread. At the moment it has no implementation, but if it ever changes, this test will verify that `attach_current_thread_permanently` contract is preserved: i.e., "permanently" means "until explicitly detached, and no through other means".

File: tests/threads_nested_attach_daemon.rs (hunk 1 above)
Comment (dmitry-timofeev) [reply tests/threads_nested_attach_daemon.rs:49]: I'd consider clarifying the intent, as in the other method, e.g.,

```
    // Dropping JNIEnv obtained from a daemon attach has no effect on the thread status.
```

**Review by dmitry-timofeev — CHANGES_REQUESTED** [review]
Thanks! The attribute and the clean ups make sense, but I think the tests shall keep these verifications, possibly, with some clarifications and SA suppressions, if needed.

The JNIEnv already has the attribute, are there any other types that need it?

File: tests/threads_nested_attach_permanently.rs (hunk 2 above)
Comment (stanislav-tkach) [reply tests/threads_nested_attach_permanently.rs:50]: The type cannot implement both `Copy` and `Clone`, so we cannot implement `Drop` without breaking changes (removing `Copy` implementation). We can silence this warning in order to preserve drop in the test, but I'm not sure if it makes sense. I guess extending the comment will be enough.

Comment (dmitry-timofeev) [reply tests/threads_nested_attach_permanently.rs:50]: I see, thanks for the explanation. Then I suggest to remove the redundant comments and assertions. @aleksuss could you do that please, or would you prefer to postpone these changes till a separate PR?

**Comment by dmitry-timofeev** [comment]
@maurolacy , big thanks for getting involved in the reviews and issue discussions, we really appreciate that! Neither Stanislav nor I currently work for companies that are jni-rs users, hence we cannot spend any company time on this project. We also no longer get any internal feedback on this project — that in Exonum was one of the biggest sources of improvements, with the Exonum team contributing numerous improvements and fixes over the years. It is very useful to have an active user involved in discussions around the project.

**Comment by maurolacy** [comment]
@dmitry-timofeev my pleasure. Contributing here was and still is a great learning opportunity for me.

**Review by dmitry-timofeev — APPROVED** [review]
LGTM, thanks!

**Pull request merged**